trappy/ftrace.py

```diff
@@ -173,74 +173,113 @@ def _trace_cache_path(self):
-        trace_metadata_path = os.path.join(cache_path, 'metadata.json')
 
         if os.path.exists(cache_path):
             shutil.rmtree(cache_path)
         os.mkdir(cache_path)
 
-        trace_metadata = {}
-
-        trace_metadata["md5sum"] = hashlib.md5(open(self.trace_path, 'rb').read()).hexdigest()
-        trace_metadata["basetime"] = self.basetime
-
-        with open(trace_metadata_path, 'w') as f:
-            json.dump(trace_metadata, f)
-
     def _update_cache(self):
         try:
             # Recreate basic cache directories only if nothing cached
             if not any([c.cached for c in self.trace_classes]):
-                self._create_trace_cache()
+                self._prepare_cache_dir()
+
+                # Write cache metadata
+                metadata_path = os.path.join(self._trace_cache_path(), 'metadata.json')
 
-            # Write out only events that weren't cached before
+                metadata = {}
+                metadata["md5sum"] = hashlib.md5(
+                    open(self.trace_path, 'rb').read()
+                ).hexdigest()
+                metadata["basetime"] = self.basetime
+                metadata["extra"] = self._get_extra_data_to_cache()
+
+                with open(metadata_path, 'w') as f:
+                    json.dump(metadata, f)
+
+            # Cache trace data
             for trace_class in self.trace_classes:
+                # Write out only events that weren't cached before
                 if trace_class.cached:
                     continue
+
                 csv_file = self._get_csv_path(trace_class)
                 trace_class.write_csv(csv_file)
         except OSError as err:
             warnings.warn(
                 "TRAPpy: Cache not created due to OS error: {0}".format(err))
 
+    def _get_extra_data_to_cache(self):
+        # By default, some amount of metadata is saved in metadata.json
+        # Additionnal metadata can be saved by overriding this method
+        return {}
+
     def _load_cache(self):
-        if self._check_trace_cache():
-            self.max_window = self._max_window()
-             # Read csv into frames
-            for trace_class in self.trace_classes:
-                try:
-                    csv_file = self._get_csv_path(trace_class)
-                    trace_class.read_csv(csv_file)
-                    trace_class.cached = True
-                except:
-                    warnstr = "TRAPpy: Couldn't read {} from cache, reading it from trace".format(trace_class)
-                    warnings.warn(warnstr)
+        cache_path = self._trace_cache_path()
+        if not os.path.exists(cache_path):
+            return
+
+        metadata = self._get_cache_metadata()
+
+        if not self._is_cache_valid(metadata):
+            shutil.rmtree(cache_path)
+            return
+
+        # Load metadata
+        self.basetime = metadata["basetime"]
+        self.max_window = self._max_window()
```

Comment:
If we're not parsing, it has to be set in `__populate_data`. As such, I would like to keep a "symmetry", in that it is not done in `_do_parse` but in each appropriate submethod. Otherwise it feels a bit confusing to me: in the scope of `_do_parse`, a newcomer would see `max_window` is only being sent after the cache is loaded (though that is my own opinion).

Reply:
Ahh yep, that does make sense actually.